mevcommitclient/mev_commit_client.go

```diff
@@ -0,0 +1,136 @@
+	MevCommitAddress           string
+	ValidatorRouterAddress     common.Address
+	BuilderRegistryAddress     common.Address
+	validatorOptInRouterCaller *validatoroptinrouter.ValidatoroptinrouterCaller
+	builderRegistryCaller      *builderRegistry.ProviderregistryCaller
+	builderRegistryFilterer    *builderRegistry.ProviderregistryFilterer
+	l1Client                   *ethclient.Client
+	mevCommitClient            *ethclient.Client
+	contractAbi                abi.ABI
+}
+
+const (
+	abiJSON = `[{"anonymous":false,"inputs":[{"indexed":true,"internalType":"address","name":"builder","type":"address"},{"indexed":false,"internalType":"uint256","name":"value","type":"uint256"},{"indexed":false,"internalType":"bytes","name":"blsPublicKey","type":"bytes"}],"name":"BuilderRegistered","type":"event"},{"inputs":[{"internalType":"address","name":"builder","type":"address"}],"name":"isBuilderValid","outputs":[],"stateMutability":"view","type":"function"}]`
+)
+
+func NewMevCommitClient(l1MainnetUrl, mevCommitUrl string, validatorRouterAddress, builderRegistryAddress common.Address) (IMevCommitClient, error) {
+	l1Client, err := ethclient.Dial(l1MainnetUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to L1 Mainnet: %w", err)
+	}
+
+	mevCommitClient, err := ethclient.Dial(mevCommitUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to MEV-Commit EVM: %w", err)
+	}
+
+	validatorOptInRouter, err := validatoroptinrouter.NewValidatoroptinrouterCaller(validatorRouterAddress, l1Client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ValidatorOptInRouter caller: %w", err)
+	}
+
+	builderRegistryCaller, err := builderRegistry.NewProviderregistryCaller(builderRegistryAddress, mevCommitClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create BuilderRegistry caller: %w", err)
+	}
+
+	builderRegistryFilterer, err := builderRegistry.NewProviderregistryFilterer(builderRegistryAddress, mevCommitClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create BuilderRegistry filterer: %w", err)
+	}
+
+	contractAbi, err := abi.JSON(strings.NewReader(abiJSON))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse contract ABI: %w", err)
+	}
+
+	return &MevCommitClient{
+		L1Address:                  l1MainnetUrl,
+		MevCommitAddress:           mevCommitUrl,
+		ValidatorRouterAddress:     validatorRouterAddress,
+		BuilderRegistryAddress:     builderRegistryAddress,
+		validatorOptInRouterCaller: validatorOptInRouter,
+		builderRegistryCaller:      builderRegistryCaller,
+		builderRegistryFilterer:    builderRegistryFilterer,
+		l1Client:                   l1Client,
+		mevCommitClient:            mevCommitClient,
+		contractAbi:                contractAbi,
+	}, nil
+}
+
+func (m *MevCommitClient) GetOptInStatusForValidators(pubkeys [][]byte) ([]bool, error) {
+	return m.validatorOptInRouterCaller.AreValidatorsOptedIn(nil, pubkeys)
+}
+func (m *MevCommitClient) GetActiveBuilders() ([][]byte, error) {
```

Comment:
How often is this function executed? Note someone could spam provider registration over time (reg, dereg, repeat), create a bunch of events, and slow down this function significantly. 

If we need to query the full set of registered providers from the mev-commit chain, a good solution may be to persist the _set of opted-in providers_ on-chain as an enumerable set -> https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/structs/EnumerableSet.sol. 

We don't use this data structure for validator registrations, because there's never a need to query the full set (just upcoming 32 proposers etc). But for providers, the set will likely be small enough that we can justify using the enum set.

Reply:
Agree with @shaspitz This will definitely cause issues if we do it everytime. One option is to have a worker constantly monitoring these events and adding to some local cache. Then we return everything from the cache.